_posts/2020-04-11-native-mobile-to-flutter.markdown

```diff
@@ -0,0 +1,122 @@
+
+We use this pattern with our SQLite database to allow sql table subscriptions that trigger redraws on subscribed user interfaces when data is modified on disk.  We also use a very simple stream subscription pattern to propagate state changes that are not tied to database state.  Flutter provides the excellent [StreamBuilder class](https://api.flutter.dev/flutter/widgets/StreamBuilder-class.html) to facilitate state propagation.
+
+We evaluated a few other patterns mentioned in flutter resources, particularly Redux and BLoC, but chose the simplicity of Scoped Model over the granularity and associated complexity that the others presented.  
+
+### Native integrations. 
+
+Flutter offers a very simple message passing api for communicating with native code.  We needed to use native integrations for a relatively small group of features: 
+
+- Push notification.
+- Deep link handling.
+- Connectivity detection.
+- Security features.
+- Device information. 
+
+_This platform specific code amounted to less than 2% of our total lines of code._ 
+
+We were able to write this functionality easily and communicate between Flutter and native with very little code overhead. Flutter provides a [very simple api](https://flutter.dev/docs/development/platform-integration/platform-channels#architecture) for communicating bidirectionally between native and dart code.
+
+### Unit tests.
+
+Unit testing in Flutter is straightforward, but lacks the robust testing utilities and third party libraries found in React.  We rolled most of our own testing scaffolding aside from the simple test runner offered by flutter.
+
+### UI testing. 
+
+UI testing in Flutter only runs in a simulator/emulator, a pretty significant drawback.  Though we were able to write effective widget tests, we found them to be brittle.  This is definitely an area that needs improvements in future iterations of Flutter.  We opted to keep the majority of presentation logic in view models so that we could test the presentation layer with unit tests wherever possible rather than employing widget tests.
+
+### Continuous integration.
+
+We opted for CodeMagic as our CI service.  It was the only commercial service we found that had specialized flutter workflows that required no additional scrpting or configuration.  We were able to integrate our git workflow with CodeMagic in under an hour.   
+
+### Security.
+
+We hired an external penetration testing firm to do a security analysis on our application when we were done.  The penetration tests found no vulnerabilities in our Flutter application.
+
+### Review of Results.
```

Comment:
These are your money quotes!

 So I'd suggest putting these first, in the intro of your article - imo they're a wonderful hook to draw the reader in. E.g.

"Ginger recently ported our native iOS and Android apps in Flutter. It was an astounding success. Our development velocity has increased by 2-3x , we now share 98% of our code across both platforms, and we've reduced the amount of code we need to maintain from 140k lines of code to just 31k. Read on to find out how we did it". 

^- That's an exaggerated (sales-y) example, but just pointing out these metrics are a great hook.

Reply:
Interesting.  Yeah, the metrics are the hook.   I'll try it with those at the top.